### ISSUE TITLE
Accuracy mismatch for ST EfficientNet LC v1 224x224x3 on Food-101

### ISSUE THREAD

**Issue by kristjansoln** [STMicroelectronics/stm32ai-modelzoo#59]
Hey!

While evaluating ST EfficientNet LC V1 (224×224×3) on Food‑101, I’m getting approx. 15 percentage points higher top‑1 accuracy with the pretrained checkpoints than what’s documented.

[Docs](https://github.com/STMicroelectronics/stm32ai-modelzoo/tree/main/image_classification/efficientnet#accuracy-with-food-101-dataset) say:
- Float: 74.84%
- Quantized: 74.44%

What I observe when I evaluate your [published pretrained weights](https://github.com/STMicroelectronics/stm32ai-modelzoo/tree/main/image_classification/efficientnet/ST_pretrainedmodel_public_dataset/food-101/st_efficientnet_lc_v1_224_tfs):
- Float (st_efficientnet_lc_v1_224_tfs.h5): 90.19%
- Quantized (st_efficientnet_lc_v1_224_tfs_int8.tflite): 89.87%

Version information:
- stm32ai-modelzoo: d1f4146
- stm32ai-modelzoo-services: f9212f7
- Tensorflow: 2.8.3

I have downloaded the Food-101 dataset using the [torchvision module](https://docs.pytorch.org/vision/stable/generated/torchvision.datasets.Food101.html).

## Training from scratch

I first trained the model from randomly initialized weights using the configuration below. This configuration should be functionally identical to [the one published for this model](https://github.com/STMicroelectronics/stm32ai-modelzoo/blob/main/image_classification/efficientnet/ST_pretrainedmodel_public_dataset/food-101/st_efficientnet_lc_v1_224_tfs/st_efficientnet_lc_v1_224_tfs_config.yaml).

```yaml
general:
  project_name: Food101-st_efficientnet_lc_v1
  logs_dir: ./output/st_efficientnet_lc_v1/Food101/logs
  saved_models_dir: ./output/st_efficientnet_lc_v1/Food101/saved_models
  display_figures: False
  global_seed: 127
  gpu_memory_limit: 10
  num_threads_tflite: 6
  model_path:

operation_mode: chain_tqe
#choices=['training' , 'evaluation', 'prediction', 'deployment', 'quantization', 'benchmarking',
#        'chain_tqeb','chain_tqe','chain_eqe','chain_qb','chain_eqeb','chain_qd ']

dataset:
  name: food-101
  training_path: /home/kristjans/envs/data/Food101/images/train/
  # validation_path:
  validation_split: 0.2
  test_path: /home/kristjans/envs/data/Food101/images/test/
  quantization_path: /home/kristjans/envs/data/Food101/images/train/
  quantization_split: 0.3

  check_image_files: False  # Optional, set it to True if you want to check that all the image files can be read successfully
  seed: 127

preprocessing:
  rescaling: { scale: 1/255.0, offset: 0.0 }
  resizing:
    interpolation: nearest
    aspect_ratio: fit
  color_mode: rgb

data_augmentation:
  random_hue:
    delta: 0.04
    change_rate: 1.0
  random_contrast:
    factor: 0.6
    change_rate: 1.0
  random_brightness:
    factor: 0.05
    change_rate: 1.0
  random_rectangle_erasing:
    nrec: (0, 10)
    area: (0.0, 0.1)
    wh_ratio: (0.25, 4.0)
    fill_method: "mosaic"
    mode: batch
    change_rate: 0.05
  random_flip:
    mode: horizontal
    change_rate: 1.0 
  random_translation:
    width_factor: 0.25
    height_factor: 0.25
    fill_mode: reflect
    interpolation: nearest
    change_rate: 1.0
  random_rotation:
    factor: 0.125
    fill_mode: reflect
    interpolation: nearest
    change_rate: 1.0
  random_zoom:
    width_factor: 0.25
    height_factor: 0.25
    fill_mode: reflect
    interpolation: nearest
    change_rate: 1.0
  random_shear:
    factor: 0.15
    fill_mode: wrap
    interpolation: nearest
    change_rate: 1.0  

training:
  model:    # Use it if you want to use a model from the zoo, mutually exclusive with 'general.model_path'
    name: st_efficientnet_lc
    version: v1
    input_shape: (224, 224, 3)
    #pretrained_weights:
    #pretrained_model_path:
  #resume_training_from:
  #frozen_layers: (0:-1)
  dropout: 0.25
  batch_size: 128
  epochs: 800
  optimizer:
    Adam:
      learning_rate: 0.0025
  callbacks:
    LRWarmupCosineDecay:
      initial_lr: 1.0e-05
      warmup_steps: 20
      max_lr: 0.0025
      hold_steps: 20
      decay_steps: 600
      end_lr: 1.0e-05
    #ReduceLROnPlateau:
    #  monitor: val_accuracy
    #  mode: max
    #  factor: 0.65
    #  patience: 20
    #  min_lr: 5.0e-05
    EarlyStopping:
      monitor: val_accuracy
      mode: max
      restore_best_weights: true
      patience: 100

quantization:
  quantizer: TFlite_converter
  quantization_type: PTQ
  quantization_input_type: uint8
  quantization_output_type: float
  export_dir: ./output/st_efficientnet_lc_v1/Food101/quantized_models

prediction:
  test_path: /home/kristjans/envs/data/Food101/images/test/ # Path to the root directory of the test set.
  
mlflow:
  uri: ./output/st_efficientnet_lc_v1/Food101/mlruns
  
hydra:
  run:
    dir: ./output/st_efficientnet_lc_v1/Food101/hydra/${now:%Y_%m_%d_%H_%M_%S}
```

The training procedure reports 74.26% accuracy on the test dataset for the float model and 73.75% for the quantized model. See the summary of the training output below. As expected, these results match the accuracies stated in the documentation.

```
2 physical GPUs, 2 logical GPUs
[INFO] : Setting upper memory limit to 5GBytes on gpu[0]
[INFO] : Setting upper limit of usable GPU memory to 5GBytes.
[INFO] : Running `chain_tqe` operation mode
[INFO] : Found 101 classes in the dataset.
[INFO] : ClearML config check
[INFO] : The random seed for this simulation is 127
Dataset stats:
  classes: 101
  training set size: 60600
  validation set size: 15150
  test set size: 25250
[INFO] : Using `st_efficientnet_lc` model
[INFO] : No pretrained weights were loaded, training from randomly initialized weights.
[INFO] : Set dropout rate to 0.25
============================================================================================================
  Model: EfficientNet-STCustom
============================================================================================================
  Layer index  Trainable    Name                         Type                      Params#  Output shape
-------------  -----------  ---------------------------  ----------------------  ---------  --------------------
            0  True         input_1                      InputLayer                      0  (None, 224, 224, 3)
            1  True         conv2d                       Conv2D                        448  (None, 112, 112, 16)

...

          279  True         top_dropout                  Dropout                         0  (None, 576)
          280  True         output_probs                 Dense                       58277  (None, 101)
------------------------------------------------------------------------------------------------------------
Total params: 571545
Trainable params:  560185
Non-trainable params:  11360
------------------------------------------------------------------------------------------------------------
Total layers: 281
Trainable layers: 281
Non-trainable layers: 0
============================================================================================================
Starting training...
Epoch 1/800
474/474 [==============================] - 241s 478ms/step - loss: 4.7367 - accuracy: 0.0096 - val_loss: 4.6521 - val_accuracy: 0.0104 - lr: 1.0000e-05
Epoch 2/800
474/474 [==============================] - 227s 479ms/step - loss: 4.6739 - accuracy: 0.0105 - val_loss: 4.6159 - val_accuracy: 0.0101 - lr: 1.3450e-04

...

Epoch 643/800
474/474 [==============================] - 224s 473ms/step - loss: 0.9546 - accuracy: 0.7306 - val_loss: 1.3728 - val_accuracy: 0.6865 - lr: 1.0000e-05
Training runtime: 1 day, 15:57:21
[INFO] : Evaluating the float model using validation_set...
119/119 [==============================] - 9s 63ms/step - loss: 1.3735 - accuracy: 0.6892
[INFO] : Accuracy of float model on validation_set = 68.92%
[INFO] : Loss of float model on validation_set = 1.3734546899795532
[INFO] : Evaluating the float model using test_set...
198/198 [==============================] - 15s 66ms/step - loss: 1.0367 - accuracy: 0.7426
[INFO] : Accuracy of float model on test_set = 74.26%
[INFO] : Loss of float model on test_set = 1.0366853475570679
[INFO] : Training complete.
[INFO] : Quantization using input images coming from /home/kristjans/envs/data/Food101/images/train/
[INFO] : Quantizing by using 30.0 % of the provided dataset...
[INFO] : Quantizing the model ... This might take few minutes ...
[INFO] : Quantization granularity : per_channel
100%|█████████████████████████████████████| 22725/22725 [15:17<00:00, 24.77it/s]
fully_quantize: 0, inference_type: 6, input_inference_type: 3, output_inference_type: 0
[INFO] : Quantization complete.
[INFO] : Evaluating the quantized model using test_set...
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
100%|███████████████████████████████████████| 198/198 [1:04:20<00:00, 19.50s/it]
[INFO] : Accuracy of quantized model on test_set = 73.75%
[INFO] : Evaluation complete.
[INFO] : chain_tqe complete.
```

## Running the evaluation for the just-trained model again

After performing the training, I ran the evaluation again for this model I just trained. I used the same configuration file, but have overridden some of the options via command line. I got identical results, as expected.

```python
float_model_path = '/home/kristjans/envs/stm/output/st_efficientnet_lc_v1/Food101/hydra/2025_07_21_16_35_00/output/st_efficientnet_lc_v1/Food101/saved_models/best_model.h5'
#float_model_path = '/home/kristjans/envs/stm/output/st_efficientnet_lc_v1/Food101/hydra/2025_07_21_16_35_00/output/st_efficientnet_lc_v1/Food101/saved_models/saved_models/best_augmented_model.h5'
quant_model_path = '/home/kristjans/envs/stm/output/st_efficientnet_lc_v1/Food101/hydra/2025_07_21_16_35_00/output/st_efficientnet_lc_v1/Food101/quantized_models/quantized_model.tflite'

output_path_root = '/home/kristjans/envs/stm/output/'
mlflow_path = output_path_root + 'st_efficientnet_lc_v1/Food101_evaluation/mlruns'
hydra_path = output_path_root +'st_efficientnet_lc_v1/Food101_evaluation/hydra/${now:%Y_%m_%d_%H_%M_%S}'
logs_dir = output_path_root +'st_efficientnet_lc_v1/Food101_evaluation/logs'

print("FLOAT MODEL EVALUATION:")

!python stm32ai-modelzoo-services/image_classification/stm32ai_main.py \
    operation_mode='evaluation' \
    general.model_path='{float_model_path}' \
    mlflow.uri='{mlflow_path}' \
    hydra.run.dir='{hydra_path}' \
    general.logs_dir='{logs_dir}'

print("QUANT MODEL EVALUATION:")

!python stm32ai-modelzoo-services/image_classification/stm32ai_main.py \
    operation_mode='evaluation' \
    general.model_path='{quant_model_path}' \
    mlflow.uri='{mlflow_path}' \
    hydra.run.dir='{hydra_path}' \
    general.logs_dir='{logs_dir}'
```

Output:

```
FLOAT MODEL EVALUATION:
2 physical GPUs, 2 logical GPUs
[INFO] : Setting upper memory limit to 10GBytes on gpu[0]
[INFO] : Setting upper limit of usable GPU memory to 10GBytes.
[INFO] : Running `evaluation` operation mode
[INFO] : Found 101 classes in the dataset.
[INFO] : ClearML config check
[INFO] : The random seed for this simulation is 127
host validation
[INFO] : Nb params of float model : 571545
[INFO] : Evaluating the float model using test_set...
198/198 [==============================] - 15s 61ms/step - loss: 1.0367 - accuracy: 0.7426
[INFO] : Accuracy of float model on test_set = 74.26%
[INFO] : Loss of float model on test_set = 1.0366853475570679
[INFO] : Evaluation complete.
QUANT MODEL EVALUATION:
2 physical GPUs, 2 logical GPUs
[INFO] : Setting upper memory limit to 10GBytes on gpu[0]
[INFO] : Setting upper limit of usable GPU memory to 10GBytes.
[INFO] : Running `evaluation` operation mode
[INFO] : Found 101 classes in the dataset.
[INFO] : ClearML config check
[INFO] : The random seed for this simulation is 127
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
host validation
[INFO] : Evaluating the quantized model using test_set...
100%|███████████████████████████████████████| 198/198 [1:01:03<00:00, 18.50s/it]
[INFO] : Accuracy of quantized model on test_set = 73.75%
[INFO] : Evaluation complete..
```

## Running the evaluation on the provided pretrained model

Finally, I ran the evaluation on the [pre-trained model, provided by the model zoo](image_classification/efficientnet/ST_pretrainedmodel_public_dataset/food-101/st_efficientnet_lc_v1_224_tfs/st_efficientnet_lc_v1_224_tfs.h5). I used the same commands/configuration as above, only changing the model path. Here, I got far better results than what was stated in the documentation, 90.1% percent accuracy for the float model and 89.87% for the quantized one.

```python
quant_model_path = '/home/kristjans/envs/stm/stm32ai-modelzoo/image_classification/efficientnet/ST_pretrainedmodel_public_dataset/food-101/st_efficientnet_lc_v1_224_tfs/st_efficientnet_lc_v1_224_tfs_int8.tflite'
float_model_path = '/home/kristjans/envs/stm/stm32ai-modelzoo/image_classification/efficientnet/ST_pretrainedmodel_public_dataset/food-101/st_efficientnet_lc_v1_224_tfs/st_efficientnet_lc_v1_224_tfs.h5'

output_path_root = '/home/kristjans/envs/stm/output/'
mlflow_path = output_path_root + 'st_efficientnet_lc_v1/Food101_evaluation/mlruns'
hydra_path = output_path_root +'st_efficientnet_lc_v1/Food101_evaluation/hydra/${now:%Y_%m_%d_%H_%M_%S}'
logs_dir = output_path_root +'st_efficientnet_lc_v1/Food101_evaluation/logs'

print("FLOAT MODEL EVALUATION:")

!python stm32ai-modelzoo-services/image_classification/stm32ai_main.py \
    operation_mode='evaluation' \
    general.model_path='{float_model_path}' \
    mlflow.uri='{mlflow_path}' \
    hydra.run.dir='{hydra_path}' \
    general.logs_dir='{logs_dir}'

print("QUANT MODEL EVALUATION:")

!python stm32ai-modelzoo-services/image_classification/stm32ai_main.py \
    operation_mode='evaluation' \
    general.model_path='{quant_model_path}' \
    mlflow.uri='{mlflow_path}' \
    hydra.run.dir='{hydra_path}' \
    general.logs_dir='{logs_dir}'
```

```
FLOAT MODEL EVALUATION:
2 physical GPUs, 2 logical GPUs
[INFO] : Setting upper memory limit to 10GBytes on gpu[0]
[INFO] : Setting upper limit of usable GPU memory to 10GBytes.
[INFO] : Running `evaluation` operation mode
[INFO] : Found 101 classes in the dataset.
[INFO] : ClearML config check
[INFO] : The random seed for this simulation is 127
host validation
[INFO] : Nb params of float model : 571545
[INFO] : Evaluating the float model using test_set...
198/198 [==============================] - 15s 61ms/step - loss: 0.3506 - accuracy: 0.9019
[INFO] : Accuracy of float model on test_set = 90.19%
[INFO] : Loss of float model on test_set = 0.3505510091781616
[INFO] : Evaluation complete.
QUANT MODEL EVALUATION:
2 physical GPUs, 2 logical GPUs
[INFO] : Setting upper memory limit to 10GBytes on gpu[0]
[INFO] : Setting upper limit of usable GPU memory to 10GBytes.
[INFO] : Running `evaluation` operation mode
[INFO] : Found 101 classes in the dataset.
[INFO] : ClearML config check
[INFO] : The random seed for this simulation is 127
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
host validation
[INFO] : Evaluating the quantized model using test_set...
100%|█████████████████████████████████████████| 198/198 [20:58<00:00,  6.36s/it]
[INFO] : Accuracy of quantized model on test_set = 89.87%
[INFO] : Evaluation complete..
```

## Question

The pre-trained model achieves 90.19 % accuracy on the test set, 15% better than what is stated in the documentation for this model. However, when training the model from scratch, the results match the documentation numbers. Also, the 90.19% accuracy exactly matches your [Flowers dataset result](https://github.com/STMicroelectronics/stm32ai-modelzoo/tree/main/image_classification/efficientnet#accuracy-with-flowers-dataset).

A 15‑point improvement is great, of course, but I’d like to be sure I’m not overlooking a bug in my configuration or something else.

Has anyone observed this behavior? Could it be possible you used a different configuration than the provided one for training your model? Or am I missing something in my training configuration or commands? 

Fantastic library, by the way.

Regards,
Kristjan

**Comment by LFOSTM** [comment]
Hello,
Thanks for your remarks.
You right, and good results are the ones in the documentation you reproduce by retraining.
You have better results evaluating the trained model because we do not have the same split training/eval/test, so when evaluating it on your side without retraining, some vectors used for the training become part of the evaluation, leading to this over estimation of the performance.
In the coming month, we will change it on our side to use the train/test of the dataset (no random split) so the user can re-evaluate the model without retraining it.
Regards,
